core/src/main/java/org/infinispan/cache/impl/CompatibilityAdvancedCache.java

```diff
@@ -0,0 +1,136 @@
+      super(cache, c -> new CompatibilityAdvancedCache<>(c, marshaller, converter), converter);
+      Collection<TypeConverter> converters = ServiceFinder.load(TypeConverter.class);
+      for (TypeConverter foundConverter : converters) {
+         // We assume these don't produce byte[] values when boxing
+         if (foundConverter.supportsInvocation(Flag.OPERATION_HOTROD)) {
+            hotRodConverter = setConverterMarshaller(new DoubleTypeConverter<>(foundConverter, converter), marshaller);
+         } else if (foundConverter.supportsInvocation(Flag.OPERATION_MEMCACHED)) {
+            memcachedConverter = setConverterMarshaller(new DoubleTypeConverter<>(foundConverter, converter), marshaller);
+         }
+      }
+      embeddedConverter = setConverterMarshaller(new DoubleTypeConverter<>(new EmbeddedTypeConverter(), converter), marshaller);
+   }
+
+   private static TypeConverter setConverterMarshaller(TypeConverter converter, Marshaller marshaller) {
+      if (marshaller != null)
+         converter.setMarshaller(marshaller);
+
+      return converter;
+   }
+
+   @Override
+   protected TypeConverter getConverter() {
+      Cache<K, V> cache = this.cache;
+      while (cache instanceof AbstractDelegatingCache && !(cache instanceof DecoratedCache)) {
+         cache = ((AbstractDelegatingCache<K, V>) cache).getDelegate();
+      }
+      if (cache instanceof DecoratedCache) {
+         EnumSet<Flag> flags = ((DecoratedCache<K, V>) cache).getFlags();
+         if (flags.contains(Flag.OPERATION_HOTROD)) {
+            return hotRodConverter;
+         } else if (flags.contains(Flag.OPERATION_MEMCACHED)) {
+            return memcachedConverter;
+         }
+      }
+      return embeddedConverter;
+   }
+
+   private static class EmbeddedTypeConverter implements TypeConverter<Object, Object, Object, Object> {
+      private static final Log log = LogFactory.getLog(EmbeddedTypeConverter.class);
+
+      private Marshaller marshaller;
+
+      @Override
+      public Object boxKey(Object key) {
+         return key;
+      }
+
+      @Override
+      public Object boxValue(Object value) {
+         return value;
+      }
+
+      @Override
+      public Object unboxKey(Object target) {
+         return unboxValue(target);
+      }
+
+      @Override
+      public Object unboxValue(Object target) {
+         if (marshaller != null && target instanceof byte[]) {
+            try {
+               return marshaller.objectFromByteBuffer((byte[]) target);
+            } catch (Exception e) {
+               throw new CacheException("Unable to unmarshall return value");
+            }
+         }
+
+         if (target instanceof byte[]) {
+            // Try standard deserialization
+            try {
+               ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream((byte[]) target));
+               return ois.readObject();
```

Comment:
This is the same as before https://github.com/infinispan/infinispan/blob/f9945e7707ca107c9ce0bfaf8808103685300d36/core/src/main/java/org/infinispan/interceptors/compat/TypeConverterInterceptor.java#L93